llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp

```diff
@@ -8178,6 +8178,87 @@ static Instruction *foldFCmpFSubIntoFCmp(FCmpInst &I, Instruction *LHSI,
   return nullptr;
 }
 
+static Instruction *foldFCmpWithFloorAndCeil(FCmpInst &I,
+                                             InstCombinerImpl &CI) {
+  Value *LHS = I.getOperand(0), *RHS = I.getOperand(1);
+  const CmpInst::Predicate Pred = I.getPredicate();
+  Type *OpType = LHS->getType();
+
+  // fcmp ole floor(x), x => fcmp ord x, 0
+  // fcmp ogt floor(x), x => false
+  if (match(LHS, m_Intrinsic<Intrinsic::floor>(m_Specific(RHS)))) {
+    if (Pred == FCmpInst::FCMP_OLE ||
+        Pred == FCmpInst::FCMP_ULE &&
+            isKnownNeverNaN(LHS, 0,
+                            CI.getSimplifyQuery().getWithInstruction(&I))) {
```

Comment:
You do not need to call isKnownNeverNaN. The only reason it was useful in instsimplify was you could not insert the new fcmp required to preserve the nan behavior

Reply:
It needs some adjustment, but you can avoid it: https://alive2.llvm.org/ce/z/YZfsRb

Reply:
Please update the alive2 link in your PR description.